bot/exts/info/patreon.py

```diff
@@ -0,0 +1,128 @@
+import datetime
+
+import arrow
+import discord
+from discord.ext import commands, tasks
+
+from bot import constants
+from bot.bot import Bot
+from bot.constants import Channels, Guild, Roles, STAFF_PARTNERS_COMMUNITY_ROLES
+from bot.decorators import in_whitelist
+from bot.log import get_logger
+from bot.utils.channel import get_or_fetch_channel
+
+log = get_logger(__name__)
+
+PATREON_INFORMATION = (
+    "Python Discord is a volunteer run non-profit organization, so we rely on Patreon donations to do what we do. "
+    "We use the money we get to offer excellent prizes for all of our events. These include t-shirts, "
+    "stickers, and sometimes even Raspberry Pis!\n\n"
+    "You can read more about how Patreon donations help us, and consider donating yourself, on our patreon page "
+    "[here](https://pydis.com/patreon)!"
+)
+
+# List of tuples containing tier number and Discord role ID.
+# Ordered from highest tier to lowest.
+PATREON_TIERS: list[tuple[int, int]] = [
+    (3, Roles.patreon_tier_3),
+    (2, Roles.patreon_tier_2),
+    (1, Roles.patreon_tier_1),
+]
+
+
+def get_patreon_tier(member: discord.Member) -> int:
+    """
+    Get the patreon tier of `member`.
+
+    A patreon tier of 0 indicates the user is not a patron.
+    """
+    for tier, role_id in PATREON_TIERS:
+        if member.get_role(role_id):
+            return tier
+    return 0
+
+
+class Patreon(commands.Cog):
+    """Cog that shows patreon supporters."""
+
+    def __init__(self, bot: Bot) -> None:
+        self.bot = bot
+
+        self.current_monthly_supporters.start()
+
+    @commands.Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member) -> None:
+        """Send a message when someone receives a patreon role."""
+        old_patreon_tier = get_patreon_tier(before)
+        new_patreon_tier = get_patreon_tier(after)
+
+        if new_patreon_tier <= old_patreon_tier:
+            return
+
+        message = (
+            f":tada: {after.mention} just became a **tier {new_patreon_tier}** patron!\n"
+            "Support us on Patreon: https://pydis.com/patreon"
+        )
+        channel = await get_or_fetch_channel(Channels.meta)
+        await channel.send(message)
+
+    async def send_current_supporters(self, channel: discord.abc.Messageable, automatic: bool = False) -> None:
+        """Send the current list of patreon supporters, sorted by tier level."""
+        guild = self.bot.get_guild(Guild.id)
+
+        embed_list = []
+        for tier, role_id in PATREON_TIERS:
+            role = guild.get_role(role_id)
+
+            # Filter out any members where this is not their highest tier.
+            patrons = [member for member in role.members if get_patreon_tier(member) == tier]
+            patron_names = [f"• {patron}" for patron in patrons]
```

Comment:
If this list is empty, imo the embed should contain something like "No patrons in this tier" as opposed to being something like this:
![image](https://user-images.githubusercontent.com/9753350/193472728-f3773292-a99b-423e-9762-f44083198fe0.png)

Reply:
That makes sense, done 👍